### PR TITLE
Integrate dns resolving

### LINKED ISSUE #571: Support for DNS addresses
## Problem statement

The [multiaddr specification](https://github.com/multiformats/multiaddr) specifies support for `dns4`, `dns6` and `dnsaddr` addresses. It is possible to [create and parse such addresses](https://github.com/status-im/nim-libp2p/blob/master/libp2p/multiaddress.nim#L407) in the Nim implementation, but there is currently no way to resolve these to dialable transport addresses. Clients trying to connect to a DNS address will encounter an [Invalid Address exception](https://github.com/status-im/nim-libp2p/blob/master/libp2p/wire.nim#L91).

DNS address resolution will be a useful feature to add for clients not using static IPs.

## Acceptance criteria

- [ ] Add a resolver for DNS multiaddrs (`dns4`, `dns6` and `dnsaddr`)

## References

- Reference resolver implementations in [js](https://github.com/multiformats/js-multiaddr/blob/master/src/resolvers/index.js) and [go](https://github.com/multiformats/go-multiaddr-dns).
- `dnsaddr` specification [here](https://github.com/multiformats/multiaddr/blob/master/protocols/DNSADDR.md)

cc @dryajov in case this has already been discussed in a different context

### PR DESCRIPTION
**Pull request by Menduist**
Follow up of #580
Closes #571

TLDR of current DNS state:
Actually resolving multiaddress can be tricky in our infra, because:
- For WSS, we need to keep the original hostname
- We would like to deduplicate resolved addresses to avoid redundant dial
- But in the case of WSS, a different hostname/ip pair can have a different meaning, and we should dial both
- We need to keep the resolving in the dialer (instead of the transport), because `dnsaddr` could resolve into different transports

So it seems that our only option is to resolve in the dialer, deduplicate on `hostname/ip` pair (or not deduplicate at all), and give the hostname to the transport.

Is there a better option?

### CODE REVIEW

**Comment by jm-clius** [comment]
Any further progress on this? 😃

**Comment by Menduist** [comment]
If this is urgent, we can merge this as is, and handle WSS dns in a future PR (since the custom host is not implemented yet in nim-websock anyway)

**Comment by jm-clius** [comment]
> If this is urgent, we can merge this as is, and handle WSS dns in a future PR (since the custom host is not implemented yet in nim-websock anyway)

Thanks, @Menduist. Not (yet) urgent, as far as I know, but will sync with team members on this. 
We _are_ using the [DNS resolver](https://github.com/status-im/nim-libp2p/pull/580) already and it works well. :)

**Comment by Menduist** [comment]
For simplicity stake, I've chosen not to deduplicate addresses for now. If it become important in the future it's still possible to add it, but as of now, that would complicate the code for few real use cases.

File: libp2p/dialer.nim [hunk 1]
```diff
@@ -54,53 +56,61 @@ proc dialAndUpgrade(
...
-            debug "Upgrade failed", msg = exc.msg, peerId
-            if exc isnot CancelledError:
-              libp2p_failed_upgrades_outgoing.inc()
-            raise exc
-
-        doAssert not isNil(conn), "connection died after upgradeOutgoing"
-        debug "Dial successful", conn, peerId = conn.peerId
-        return conn
+  for rawAddress in addrs:      # for each address
+    address = rawAddress
```
Comment (dryajov) [inline libp2p/dialer.nim:60]: It doesn't look like we need `var address` from above anymore and we should be able to simplify the for to `for address in addrs: ...`?

File: tests/testnameresolve.nim [hunk 2]
```diff
@@ -90,18 +90,6 @@ suite "Name resolving":
 
       check testOne("/ip6/::1/tcp/0", "/ip6/::1/tcp/0")
 
-    asyncTest "test multiple resolve":
```
Comment (dryajov) [inline tests/testnameresolve.nim:93]: why is this removed?

Comment (dryajov) [reply tests/testnameresolve.nim:93]: oh I see, you're removing `resolveMAddresses`... hmm, ok I guess it is a better interface overall since a name can resolve to multiple addrs.

File: tests/testswitch.nim [hunk 3]
```diff
@@ -901,5 +901,28 @@ suite "Switch":
       switch1.peerStore.addressBook.get(switch2.peerInfo.peerId) == switch2.peerInfo.addrs.toHashSet()
       switch2.peerStore.addressBook.get(switch1.peerInfo.peerId) == switch1.peerInfo.addrs.toHashSet()
 
-      switch1.peerStore.addressBook.get(switch2.peerInfo.peerId) == switch2.peerInfo.addrs.toHashSet()
-      switch2.peerStore.addressBook.get(switch1.peerInfo.peerId) == switch1.peerInfo.addrs.toHashSet()
+      switch1.peerStore.protoBook.get(switch2.peerInfo.peerId) == switch2.peerInfo.protocols.toHashSet()
+      switch2.peerStore.protoBook.get(switch1.peerInfo.peerId) == switch1.peerInfo.protocols.toHashSet()
+
+  asyncTest "e2e dial dns address":
```
Comment (dryajov) [inline tests/testswitch.nim:907]: I think we're missing tests for hostnames that resolve to multiple addrs, it would also be nice to test this across both tcp and ws transports.

File: libp2p/dialer.nim (hunk 1 above)
Comment (Menduist) [reply libp2p/dialer.nim:60]: Good point, this code wasn't working in 1.4 actually, I fixed it

File: tests/testnameresolve.nim (hunk 2 above)
Comment (Menduist) [reply tests/testnameresolve.nim:93]: Yeah, we could wrap `resolveMAddress` in a `resolveMAddresses`, but we wouldn't use it in libp2p since we need to keep the original hostname for each original maddress

File: tests/testswitch.nim (hunk 3 above)
Comment (Menduist) [reply tests/testswitch.nim]: Done

**Comment by Menduist** [comment]
While adding the last test, I though about an other complication

If we have
`/dnsaddr/test1.io` which resolves to `/dns4/test2.io/tcp/64/wss`

We would ideally need to dial test2.io using the test2.io hostname, but currently we keep the original test1.io hostname, which could create issues while using wss

(to be fair, right now it wouldn't work at all since we don't allow resolving dns4 addresses inside a dnsaddr)

Someone said can of worms? :sweat_smile:
I'll leave it as is for now, but something to keep in mind..

**Comment by dryajov** [comment]
> Maybe a way to handle all of this more cleanly would be to split the dnsaddr and dns* resolution in two phases, dnsaddr in the dialer and dns* in the transport

Actually, this doesn't sound like a bad idea, but I need to think about it a little more - can of worms 😭

**Review by dryajov — APPROVED** [review]
LGTM, even tho there might be dragons.
